static/js/lazy_set.ts

```diff
@@ -0,0 +1,112 @@
+import * as blueslip from "./blueslip";
+
+export class LazySet {
+    private data:
+        | {
+              arr: number[];
+              set: undefined;
+          }
+        | {
+              arr: undefined;
+              set: Set<number>;
+          };
+
+    /*
+        This class is optimized for a very
+        particular use case.
+
+        We often have lots of subscribers on
+        a stream.  We get an array from the
+        backend, because it's JSON.
+
+        Often the only operation we need
+        on subscribers is to get the length,
+        which is plenty cheap as an array.
+
+        Making an array from a set is cheap
+        for one stream, but it's expensive
+        for all N streams at page load.
+
+        Once somebody does an operation
+        where sets are useful, such
+        as has/add/delete, we convert it over
+        to a set for a one-time cost.
+    */
+
+    constructor(vals: number[]) {
+        this.data = {
+            arr: vals,
+            set: undefined,
+        };
+    }
+
+    get size(): number {
+        const {data} = this;
+        if (data.set !== undefined) {
+            return data.set.size;
+        }
+
+        return data.arr.length;
+    }
+
+    keys(): IterableIterator<number> {
+        const {data} = this;
+        if (data.set !== undefined) {
+            return data.set.keys();
+        }
+        return data.arr.values();
+    }
+
+    _make_set(): void {
+        if (this.data.set !== undefined) {
+            return;
+        }
+
+        this.data = {
+            set: new Set(this.data.arr),
+            arr: undefined,
+        };
+    }
+
+    map<T>(f: (v: number, k: number) => T): T[] {
+        return Array.from(this.keys(), f);
+    }
+
+    has(v: number): boolean {
+        this._make_set();
+        if (this.data.set === undefined) {
+            throw new Error("_make_set() did not convert the array into a set.");
+        }
```

Comment:
This error case could be avoided by having `_make_set` return the `Set`.

Reply:
Sorry for the delay. Done.